File: 0054-Spiral Matrix/0054-Spiral Matrix.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        vector<int> result;
        int m = matrix.size();
        if (m == 0) {
            return result;
        }
        int n = matrix[0].size();
        int rowStart = 0, rowEnd = m - 1;
        int colStart = 0, colEnd = n - 1;
        while (rowStart <= rowEnd && colStart <= colEnd) {
            for (int j = colStart; j <= colEnd; ++j) {
                result.push_back(matrix[rowStart][j]);
            }
            ++rowStart;
            
            for (int i = rowStart; i <= rowEnd; ++i) {
                result.push_back(matrix[i][colEnd]);
            }
            --colEnd;
            
            if (rowStart <= rowEnd) {
                for (int j = colEnd; j >= colStart; --j) {
                    result.push_back(matrix[rowEnd][j]);
                }
                --rowEnd;
            }
            
            if (colStart <= colEnd) {
                for (int i = rowEnd; i >= rowStart; --i) {
                    result.push_back(matrix[i][colStart]);
                }
                ++colStart;
            }
        }
        
        return result;
    }
};

File: 0001-0100/0054-Spiral Matrix/0054-Spiral Matrix.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        m = len(matrix)
        if m == 0:
            return []
        n = len(matrix[0])
        result = []
        rowStart = 0
        rowEnd = m - 1
        colStart = 0
        colEnd = n - 1
        while rowStart <= rowEnd and colStart <= colEnd:
            for j in range(colStart, colEnd + 1):
                result.append(matrix[rowStart][j])
            rowStart += 1
            for i in range(rowStart, rowEnd + 1):
                result.append(matrix[i][colEnd])
            colEnd -= 1
            if rowStart <= rowEnd:
                for j in range(colEnd, colStart - 1, -1):
                    result.append(matrix[rowEnd][j])
                rowEnd -= 1
                
            if colStart <= colEnd:
                for i in range(rowEnd, rowStart - 1, -1):
                    result.append(matrix[i][colStart])
                colStart += 1
        return result
